[draft] ffmpeg chapters

kode54 opened the pull request:
Here is a draft, showing it's possible to read chapters. With some work, one could even use FFmpeg to write tags by altering the tag dictionaries, then transmuxing to a new file, closing, then moving the new file into place.

There are some workarounds added for now, which allow the tag reader to take priority, or else TagLib will prevent the chapters from being seen.

I'm open to any further work this might need, but perhaps you could take a stab at redoing parts of it.

tmtom commented:
Thanks for you work, was able to rebase on master and cleanly build on Ubuntu 24.04.
I believe this would resolve https://github.com/fooyin/fooyin/issues/694, right?
Did quick check on a m4a with chapters but did not worked for me. Tried m4a create by https://github.com/yt-dlp/yt-dlp
using `-f bestaudio --format-sort acodec:m4a --embed-thumbnail --embed-metadata --add-metadata` and downloaindg random YT video with chapters (this creates m4a with chapters and cover art) but still shows as a single track in fooyin.

kode54 commented:
Unfortunately, I don't know how it caches track info. It has this annoying behavior where any other Reader class will take it over and present a single track. I tried to override it so it places the FFmpeg Reader as first.

Unfortunately, this also conflicts with tag writing for the supported formats. An FFmpeg tag writer will need to be added to this.

Edit: Oh, you can reorder them yourself. You'll need to drag the FFmpeg tag reader above the TagLib one.

ludouzi commented:
Thanks for working on this.

I’ll take another look at #882 first (and make the suggested improvements), then give this a test.

kode54 committed:
[core,ffmpeg] Implement chapter support

v2: Rebase, minor tweaks
v3: Formatting

kode54 committed:
[core,ffmpeg] Better tag reading

v2: Rebase, minor tweaks

kode54 committed:
[core,ffmpeg] Adjust seeking a bit

Some formats don't have very accurate packet seeking, so
possibly seek a little early, then skip ahead using PTS, like
APE already did.

v2: Handle PTS creating invalid audio buffer timestamps
v3: Seeking hack for chapters
v4: Rebase, minor tweaks
